More complete default for template_txt for service.mqtt

michlv opened the pull request.

jim-easterbrook commented:
Won't this break existing users' setups by renaming things, e.g. `wind_dir` is renamed `wind_dir_degrees`?

michlv commented:
> 
> 
> Won't this break existing users' setups by renaming things, e.g. `wind_dir` is renamed `wind_dir_degrees`?

Hi Jim, not it will not. This is only used as default in weather.ini file if template_txt is not specified. Even then, when pywws stops and template_txt was not specified, pywws will write new version of weather.ini defining the template_txt with the current version. Practically this means that nobody has weather.ini without template_txt. So it really only helps new people.

I think that this new version is much better default. I tried to connect pywws to Domoticz and it took me a while to come up with this default, which provides enough info. If it was like this right from start, it would have taken me much less time to integrate.

jim-easterbrook commented:
Thanks, I hadn't really thought it through properly. I'll merge in a day or two if none of the MQTT gurus (@GeekyTim @sgtwilko @matt2005 @mortenf @rk295) has any objection.

sgtwilko commented:
My only comment would be that this change should be documented.

If I were setting up a new weather station with the intent of using an existing MQTT setup it wouldn't be obvious why two setups on the same version behave differently.

jim-easterbrook commented:
Could the original names be included as well? I notice that illuminance and UV values are in the new default, so most users (i.e. the ones without solar sensors) will be forced to edit their template anyway.

michlv commented:
With regards of setting up new weather station. I would think that this highly depends if the user already used pywws or not. I would think that if that is existing user, one would copy over weather.ini from the other installation and amend it, in which case there is no issue, because weather.ini will have the original template_txt. These new defaults will only affect new users, who are setting this up the first time by using documentation.

michlv commented:
With regards of the uv and illuminance, I thought that if one uses the below format:

#illuminance  \\'"illuminance_lux" : "%.1f",\\'#

the value only gets output if the weather station supports it. Otherwise nothing (at least https://pywws.readthedocs.io/en/latest/api/pywws.template.html#module-pywws.template for the #key suggest it.). So I think the MQTT message will not have anything for illuminance or uv, if the weather station model does not support those.

I have not actually tested it with pretending that my weather station does not support these values. I'll give it a try.

michlv commented:
And documenting the change, I have put a line into the change log. Is that what you had in mind sgtwilko?

jim-easterbrook commented:
Stations without solar sensors will most likely raise a KeyError exception if you try to use solar data. This is why other service modules, such as underground, add solar data to the template in the `__init__` method.
Missing data values, for example when the station briefly loses contact with its outside sensors, are given the value `None` and can be omitted or represented by '-' or whatever in a template as required.

michlv commented:
Thanks Jim, I think about this and come back hopefully with workable solution :-)

sgtwilko commented:
@michlv I was thinking it should also be in the official documents on readthedocs.io

Just something in the MQTT section mentioning the default template has breaking changes from version X.

jim-easterbrook commented:
There's a standard in the Sphinx markup for this. I'd have to look it up but it results in a section starting with something like "New in version vNN.N:".

GeekyTim commented:
> Thanks, I hadn't really thought it through properly. I'll merge in a day or two if none of the MQTT gurus (@GeekyTim @sgtwilko @matt2005 @mortenf @rk295) has any objection.

No objection for me!  Currently working on my weather clock v2 where I'm taking MQTT dat from pywws and displaying on a 64x64 LED matrix.  These updates are always welcome!

michlv commented:
Hopefully I have addressed all comments now.
* Added versionchanged comment in MQTT for template_txt (just not sure what version to put there).
* The template adjusts itself based on type of weather station.
* Changed the documentation to use the default template setting in example, but still left the customization in place as separate point.
* Changed formatting of template_txt in weather.ini when defaulted. Previously it was all on one line, now it formats the same as stated in documentation. This will make it easier to edit by human.

jim-easterbrook commented on src/pywws/service/mqtt.py:
```diff
@@ -147,35 +179,65 @@ class ToService(pywws.service.LiveDataService):
+            self.template += self.template_3080_add
         # get template text
         template = literal_eval(context.params.get(
-            service_name, 'template_txt', pprint.pformat(self.template)))
```
What was wrong with pprint.pformat?

michlv replied:
the whole template ended up on one line, if defaulted, i.e.:

template_txt = '\n#idx          \'"idx"             : "%Y-%m-%d %H:%M:%S",\'#\n#wind_dir     \'"wind_dir_degrees": "%.d",\' \'\' \'winddir_degrees(x)\'#\n#wind_dir     \'"wind_dir_text"   : "%s",\' \'\' \'winddir_text(x)\'#'

which makes it difficult to change. The new version will format it as (in the .ini file):

template_txt = ('#idx          \'"idx"             : "%Y-%m-%d %H:%M:%S",\'#\n'
    '#wind_dir     \'"wind_dir_degrees": "%.d",\' \'\' \'winddir_degrees(x)\'#\n'
    '#wind_dir     \'"wind_dir_text"   : "%s",\' \'\' \'winddir_text(x)\'#'
    )

that is format the documentation uses. This is one value per line and also follows the format used in the source code to specify default.

I removed the template_txt from the example in documentation, to encourage people to use the default and make it easier to start (example is smaller), but then felt that if the resulting format of defaulted template_txt was difficult to change, users may not be happy about it. This way it actually makes it easier, default example is small, template_txt gets defaulted for you, and if any changes are needed later (new values or just cleanup) one can do it easier this way.

pprint.format is still used, but for each line separately.

jim-easterbrook replied:
OK, I thought pprint had given me the right thing when I tried it, but I might have misremembered. I wonder if it can be configured to give one item per line, which I agree is what we want.

michlv replied:
I checked but did not see anything there. From python perspective it is just one string and we want the content formatted special way, taking each line, escaping it and preserving the lines in both escaped string and the output once it is processed through literal_eval.

michlv commented on src/pywws/service/mqtt.py:
```diff
@@ -92,9 +116,17 @@
+default will be used, which sends a lot of values in metric and imperial
+units.
+
+.. versionchanged:: NN.N
```
Please let me know what version should  I put here or if you are happy to update it once merged?

jim-easterbrook replied:
I'll set it when I do the next release. That's when I normally modify the changelog, using GitHub's "changes since the last release" to remind me what to say.
